sys/cheri/cheric.h

```diff
@@ -123,49 +119,21 @@ cheri_is_address_inbounds(const void * __capability cap, vaddr_t addr)
  * appears not currently to be the case, so manually derive using
  * cheri_getpcc() for now.
  */
-static __inline void * __capability
-cheri_codeptr(const void *ptr, size_t len)
-{
-#ifdef NOTYET
-	void (* __capability c)(void) = ptr;
-#else
-	void * __capability c = cheri_setaddress(cheri_getpcc(),
-	    (register_t)ptr);
-#endif
-
-	/* Assume CFromPtr without base set, availability of CSetBounds. */
-	return (cheri_setbounds(c, len));
-}
-
-static __inline void * __capability
-cheri_codeptrperm(const void *ptr, size_t len, register_t perm)
-{
-
-	return (cheri_andperm(cheri_codeptr(ptr, len),
-	    perm | CHERI_PERM_GLOBAL));
-}
-
-static __inline void * __capability
-cheri_ptr(const void *ptr, size_t len)
-{
-
-	/* Assume CFromPtr without base set, availability of CSetBounds. */
-	return (cheri_setbounds((__cheri_tocap const void * __capability)ptr, len));
-}
+#define cheri_codeptr(ptr, len)	\
+	cheri_setbounds(__builtin_cheri_cap_from_pointer(cheri_getpcc(), ptr), len));
 
-static __inline void * __capability
-cheri_ptrperm(const void *ptr, size_t len, register_t perm)
-{
+#define cheri_codeptrperm(ptr, len, perm)	\
+	cheri_andperm(cheri_codeptr(ptr, len), perm | CHERI_PERM_GLOBAL)
 
-	return (cheri_andperm(cheri_ptr(ptr, len), perm | CHERI_PERM_GLOBAL));
-}
+#define cheri_ptr(ptr, len)	\
+	cheri_setbounds(    \
+	    (__cheri_tocap __typeof__((ptr)[0]) *__capability)ptr, len)
```

Comment:
Can you use `(__cheri_tocap __typeof((ptr)) __capability)`?

Reply:
Didn't seem to work with array decay:
`error: invalid __cheri_tocap from 'int *' to unrelated type 'typeof ((array)) * __capability' (aka 'int (* __capability)[3]')`

I don't really feel like fixing this in the compiler right now since the workaround with [0] works.

Reply:
Ok, no worries.  I was able to use that for PTR2CAP(), but that might not be used with arrays.

Reply:
I blame clang-format 😇

Reply:
Presumably it thinks __capability is the variable name?..

Reply:
Yeah, clang-format is dumb, it thinks the following is the right formatting:
```
(void *__capabiliity) & foo;
```
I.e it thinks it's declaring a `void *` called `__capability`, then doing a bitwise `&` of that with `foo`. Our clang-format does the same too; we might want to fix that.